packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp

```diff
@@ -0,0 +1,251 @@
+      regX[j]->elementWiseMultiply(damping, *diag_inv[j], *regRes[j], SC_ONE);
+    }
+  }
+
+  return;
+} // jacobiIterate
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void smootherSetup(RCP<Teuchos::ParameterList> params,
+                   const int maxRegPerProc,
+                   const std::vector<RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > > revisedRowMapPerGrp,
+                   const std::vector<RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionGrpMats,
+                   const std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionInterfaceScaling) {
+  const std::string type = params->get<std::string>("smoother: type");
+
+  std::map<std::string, int> smootherTypes;
+  smootherTypes.insert(std::pair<std::string, int>("None",      0));
+  smootherTypes.insert(std::pair<std::string, int>("Jacobi",    1));
+  smootherTypes.insert(std::pair<std::string, int>("Gauss",     2));
+  smootherTypes.insert(std::pair<std::string, int>("Chebyshev", 3));
+
+  switch(smootherTypes[type]) {
+  case 0:
+    break;
+  case 1:
+    jacobiSetup(params, maxRegPerProc, revisedRowMapPerGrp, regionGrpMats, regionInterfaceScaling);
+  case 2:
+    std::cout << "Gauss-Seidel smoother not implemented yet no smoother is applied" << std::endl;
+    break;
+  case 3:
+    std::cout << "Chebyshev smoother not implemented yet no smoother is applied" << std::endl;
+    break;
+  }
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void smootherApply(RCP<Teuchos::ParameterList> params,
+                   const int maxRegPerProc,
+                   std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > >& regX,
+                   const std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regB,
+                   const std::vector<RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionGrpMats,
+                   const std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionInterfaceScaling,
+                   const RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > mapComp,
+                   const std::vector<RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > > rowMapPerGrp,
+                   const std::vector<RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > > revisedRowMapPerGrp,
+                   const std::vector<RCP<Xpetra::Import<LocalOrdinal, GlobalOrdinal, Node> > > rowImportPerGrp) {
+  const std::string type = params->get<std::string>("smoother: type");
+
+  std::map<std::string, int> smootherTypes;
+  smootherTypes.insert(std::pair<std::string, int>("None",         0));
+  smootherTypes.insert(std::pair<std::string, int>("Jacobi",       1));
+  smootherTypes.insert(std::pair<std::string, int>("Gauss-Seidel", 2));
+  smootherTypes.insert(std::pair<std::string, int>("Chebyshev",    3));
+
+  switch(smootherTypes[type]) {
```

Comment:
@lucbv Since there's no input parameter validation for now, it might make sense to add something along the lines of 
```
default:
{
  // --> throw error stating that this smoother type is unkwown.
  break;
}

Reply:
Yeah I thought about it, although at the moment it will just not perform any smoothing which mean the solver simply won't converge after the first v-cycle.
I don't know that I want to spend more time on this know since I set Jacobi as default in the parameter list and it is the only supported option so far...

Reply:
It's fine. Maybe leave this for any future improvements.